docs/architecture.md

```diff
@@ -1,32 +1,139 @@
+
+#### Load balancing across distributors
+
+We recommend randomly load balancing write requests across distributor instances, potentially by running the distributors as a Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/).
+
+### Ingester
+
+The **ingester** service is responsible for writing sample data to long-term storage backends (DynamoDB, S3, Cassandra, etc.).
+
+#### Shared-nothing processes
+
+Ingesters are, for the most part, shared-nothing processes. They don't coordinate or communicate with each other *under normal operation*. The exception is on rolling updates, during which ingesters transfer their in-memory data from the ingester exiting the group to the ingester joining the group.
+
+#### State
+
+Ingesters are semi-stateful in that they always retain the last 12 hours worth of samples. When restarting or upgrading ingesters, care must be taken to avoid losing that data.
+
+As *semi*-stateful processes, ingesters are *not* designed to be long-term data stores. In Cortex, that role is played by the [chunk store](#chunk-store).
+
+#### Write de-amplification
+
+Ingesters semi-statefully store the last 12 hours worth of samples in order to perform **write de-amplification**, i.e. batching and compressing samples for the same series and flushing them out to the [chunk store](#chunk-store). Under normal operations, there should be *many* orders of magnitude fewer queries per second (QPS) worth of writes to the chunk store than to the ingesters.
+
+This division of labor between the ingester and the chunk store is the main source of Cortex's low total cost of ownership (TCO).
+
+### Querier
+
+The **querier** service handles the actual [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) evaluation of samples stored in long-term storage.
+
+It embeds the chunk store client code for fetching data from long-term storage and communicates with [ingesters](#ingester) for more recent data.
+
+### Ruler
+
+The **ruler** service is responsible for handling alerts produced by [Alertmanager](https://prometheus.io/docs/alerting/alertmanager/).
+
+### Query frontend
+
+The **query frontend** is an optional service that accepts HTTP requests, queues them by tenant ID, and retries in case of errors.
+
+> As mentioned [above](#distributor), the query frontend is completely optional; you can use queriers directly. To use the query frontend, direct incoming authenticated traffic at them and set the `-querier.frontend-address` flag on the queriers.
+
+#### Queueing
+
+Queuing performs a number of functions for the query frontend:
+
+* It ensures that any large queries don't case an out-of-memory (OOM) error in the querier. This allows administrators to over-provision querier parallelism.
```

Comment:
s/case/cause/

Its more subtle than that though; queries can OOM a querier, but the query frontend service will retry those that do.  This allows you to under provision memory for your queries (or optimistically run more small queries in parallel) helping reduce TCO.